src/core/algorithms/gfd/gfd_miner.cpp

```diff
@@ -0,0 +1,660 @@
+                Embedding const& embedding) {
+    if (token.first == -1) {
+        tok = token.second;
+    } else {
+        vertex_t u = boost::vertex(token.first, pattern);
+        vertex_t v = embedding.at(u);
+        std::map<std::string, std::string> attrs = graph[v].attributes;
+        if (attrs.find(token.second) == attrs.end()) {
+            return false;
+        }
+        tok = attrs.at(token.second);
+    }
+    return true;
+}
+
+bool Satisfied(std::vector<Literal> const& literals, graph_t const& graph, graph_t const& pattern,
+               Embedding const& embedding) {
+    for (Literal const& l : literals) {
+        Token fst_token = l.first;
+        Token snd_token = l.second;
+        std::string fst, snd;
+        bool first_res = CheckToken(fst_token, fst, graph, pattern, embedding);
+        bool second_res = CheckToken(snd_token, snd, graph, pattern, embedding);
+        if (!first_res or !second_res or fst != snd) return false;
+    }
+    return true;
+}
+
+bool Validate(graph_t const& graph, Gfd const& gfd, Embeddings const& embeddings,
+              std::size_t const sigma) {
+    bool result = false;
+    std::map<vertex_t, std::set<vertex_t>> maps = {};
+    for (auto& embedding : embeddings) {
+        if (!Satisfied(gfd.GetPremises(), graph, gfd.GetPattern(), embedding)) {
+            continue;
+        }
+        result = true;
+
+        if (!Satisfied(gfd.GetConclusion(), graph, gfd.GetPattern(), embedding)) {
+            return false;
+        }
+
+        BGL_FORALL_VERTICES_T(v, gfd.GetPattern(), graph_t) {
+            if (maps.find(v) != maps.end()) {
+                maps.at(v).insert(embedding.at(v));
+            } else {
+                std::set<vertex_t> vertices = {embedding.at(v)};
+                maps.emplace(v, vertices);
+            }
+        }
+    }
+
+    std::size_t min_size = boost::num_vertices(graph);
+    for (auto& [v, image] : maps) {
+        min_size = std::min(image.size(), min_size);
+    }
+    return result && (min_size >= sigma);
+}
+
+std::vector<Literal> GenerateLiterals(graph_t const& pattern,
+                                      std::map<std::string, Info> const& attrs_info) {
+    std::vector<size_t> indices(boost::num_vertices(pattern));
+    std::iota(std::begin(indices), std::end(indices), 0);
+    std::vector<std::vector<size_t>> pairs = GetSubsets<size_t>(indices, 2);
+
+    std::vector<Literal> result = {};
+    for (auto& index_pair : pairs) {
+        size_t fst = index_pair.at(0);
+        size_t snd = index_pair.at(1);
+        std::string fst_label = pattern[boost::vertex(fst, pattern)].attributes.at("label");
+        std::string snd_label = pattern[boost::vertex(snd, pattern)].attributes.at("label");
+
+        for (auto& fst_name : attrs_info.at(fst_label)) {
+            Token fst_token = Token(fst, fst_name.first);
+            for (auto& snd_name : attrs_info.at(snd_label)) {
+                Token snd_token = Token(snd, snd_name.first);
```

Comment:
Applies everywhere else
```suggestion
                Token snd_token = {snd, snd_name.first};
```